It defines an IsAuthenticated decorator that is applied to routes.

victor-shelepen opened the pull request:
IsAuthenticated decorator has been created that is applied to route actions.

## Description
There is functionality that provides authorization functionality. There are examples in imperative way these show how to use it. But it likes configurable rules. Principal has also to provide its functionality in a declarative way by decorators. I have not also found an example of decorator usage  

## Related Issue
[inversify-express-utils - An example of a custom decorator. Cover the Principal functionality with decorators.](https://github.com/inversify/inversify-express-utils/issues/454)

## Motivation and Context
It will provide the functionality in a declarative style. The project will provide functionality by configurable rules. It is convenient and structural approach.

## How Has This Been Tested?
## Expected Behavior
```
  @httpGet('/acl/:_id')
  @permission('fetch_any_acl')
  private async acl (request: Request) {
    const hasAccess = await this.httpContext.user.isResourceOwner('fetch_any_acl')
    if (!hasAccess) {
      return this.json({ message: 'Forbidden' }, 403)
    }
    const userData = await this.userEntity.getFull(new ObjectId(request.params._id))

    return this.json(userData, 200)
  }
```
## Current Behavior
```
  @httpGet('/acl/:_id')
  @permission('fetch_any_acl')
  private async acl (request: Request) {
    const userData = await this.userEntity.getFull(new ObjectId(request.params._id))

    return this.json(userData, 200)
  }
```

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

Commit:
It defines an IsAuthenticated decorator that is applied to routes.

vlikin commented on test/auth_provider.test.ts:
```diff
@@ -19,7 +21,7 @@ describe("AuthProvider", () => {
     });
 
-    it("Should be able to access current user via HttpContext", (done) => {
+    xit("Should be able to access current user via HttpContext", (done) => {
```
I see. I've disabled another test.

Commit:
Small fix. The disable test has been enabled.

Commit:
Lint fixing.

vlikin commented on test/auth_provider.test.ts:
```diff
@@ -70,8 +72,8 @@ describe("AuthProvider", () => {
                 if (this.httpContext.user !== null) {
                     const email = this.httpContext.user.details.email;
                     const name = this._someDependency.name;
-                    const isAuthenticated = await this.httpContext.user.isAuthenticated();
```
I did it because Travis had failed because a lint error: shadow variable.

Commit:
All decorators have been generated for Principal.

Access logic has been separated into a decorator factory.
Decorators remained have been implemented using the factory.

dcavanagh commented:
thanks for the PR. Could you please update the documentation to explain the use case and how to use it.

Thank you

vlikin commented:
@dcavanagh Yes, I will do. :)

Commit:
Authentication provider decorators documentation. Also some fixes.

* Decorators are exportable from the index declaration file.

vlikin commented:
@dcavanagh I've added the documentation. It is available [here](https://github.com/vlikin/inversify-express-utils/tree/vlikin-inversify-express-principle-decorators#decorators-1). The decorators are applied only on actions now. We can also extend the functionality for controllers. This is a good idea to disclose how to implement custom decorators for controllers and actions  in documentation.

PodaruDragos commented:
will this be merged ?

PodaruDragos commented on src/decorators.ts:
```diff
@@ -130,3 +131,50 @@ export function params(type: PARAMETER_TYPE, parameterName?: string) {
+
+export function isAuthenticated (pass = true): any {
+    return httpContextAccessDecoratorFactory(async (context) => {
+        return (await context.user.isAuthenticated() && pass);
```
why do we need `&& pass` here ?

PodaruDragos commented on README.md:
````diff
@@ -513,6 +513,49 @@ class UserDetailsController extends BaseHttpController {
 ```
 
+### Decorators
+There are helpers for AuthProvider these are wrapped into decorators these are applied to action methods.
````
this English seems a bit shady, I am not a native speaker myself so can't really help with correct phrasing

PodaruDragos commented:
hey @victor-shelepen if you can get this green with master and address some of the comments I'll happily merge this

PodaruDragos commented:
> @dcavanagh I've added the documentation. It is available [here](https://github.com/vlikin/inversify-express-utils/tree/vlikin-inversify-express-principle-decorators#decorators-1). The decorators are applied only on actions now. We can also extend the functionality for controllers. This is a good idea to disclose how to implement custom decorators for controllers and actions in documentation.

Will be great if we can have this for controllers as well

victor-shelepen commented:
> hey @victor-shelepen if you can get this green with master and address some of the comments I'll happily merge this

:) Thank you. I agree.
